src/daft-scan/src/scan_task_iters.rs

```diff
@@ -0,0 +1,78 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+
+use crate::{ScanTask, ScanTaskRef};
+
+type BoxScanTaskIter = Box<dyn Iterator<Item = DaftResult<ScanTaskRef>>>;
+
+/// Coalesces ScanTasks by their filesizes
+pub fn merge_by_sizes(scan_tasks: BoxScanTaskIter, target_filesize: usize) -> BoxScanTaskIter {
+    Box::new(MergeByFileSize {
+        iter: scan_tasks,
+        target_filesize,
+        buffer: std::default::Default::default(),
+    })
+}
+
+struct MergeByFileSize {
+    iter: BoxScanTaskIter,
+    target_filesize: usize,
+
+    // HACK: we use a Vec instead of IndexMap because PartitionSpec is not Hashable
+    // This might potentially be very slow. We should explore another option here?
+    // Could be easier to make `PartitionSpec: Hash` if we don't use a `Table` under the hood.
+    buffer: Vec<ScanTaskRef>,
```

Comment:
Very iffy still about this choice - please leave comments here on how this can be improved

Ideally, I'd like to maintain a buffer which is a mapping between `{(PartitionSpec, FileFormatConfig, Schema, StorageConfig, Pushdown): <current merged ScanTask with filesize < 128MB>}`.

However, not all of these implement `Hash`, and in particular `PartitionSpec` is currently backed by a `Table` which might be tricky to implement `Hash`. Hence I opted for this `Vec` approach and iteratively check equality... but this seems very slow.

Reply:
Resolved -- changed to use a single-element accumulator instead